### ISSUE TITLE
Task to AsyncResult transition wraps `Exception` into the `AggregateException`

### ISSUE THREAD

**Issue by xperiandri** [demystifyfp/FsToolkit.ErrorHandling#154]
**Describe the bug**
```
    member this.AsyncAcquireTokenInteractively (uiThread, correlationId) = asyncResult {
        do! Async.SwitchToContext uiThread
        let builder =
            pca.AcquireTokenInteractive(options.Scopes)
                .WithCorrelationId(correlationId)
        if not (isNull configureInteractive) then configureInteractive.Invoke(builder)
        try return! builder.ExecuteAsync() // FSharp.Control.FusionTasks is used not to write |> Async.AwaitTask
        with :? MsalClientException as ex -> return! Error ex.ErrorCode
    }
```
Matching by the `MsalClientException` does not happen as it is wrapped by the `AggregateException`.

**To Reproduce**
1. Call Task returning method via `return!` in an `asyncResult` CE
2. Catch exception
3. You will see that it is always the `AggregateException`.

**Expected behavior**

**Desktop (please complete the following information):**
 - Version 2.10.0

**Comment by TheAngryByrd** [comment]
I don't think this is a problem with FsToolkit as this is how exceptions are throw with Tasks via dotnet https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/exception-handling-task-parallel-library

**Comment by xperiandri** [comment]
Is it how `async` CE works itself?

**Comment by xperiandri** [comment]
It does not happen with `await` in C#

**Comment by xperiandri** [comment]
Ah, this is the issue https://github.com/fsharp/fslang-suggestions/issues/840

**Comment by xperiandri** [comment]
As long as it is a FSharp.Core issue I propose to use a corrected way of awaiting tasks
https://github.com/jet/FsKafka/blob/6c6186f822a221adf0ad06e8a44a392aa55aa8ec/src/FsKafka/Infrastructure.fs#L14 instead of `Async.AwaitTask`

**Comment by TheAngryByrd** [comment]
Hey @NinoFloris see any downsides to doing this?

**Comment by njlr** [comment]
I have this helper in various code-bases:

```fsharp
let rec unpackException (exn : Exception) = 
  match exn with
  | :? AggregateException as agg -> 
    match Seq.tryExactlyOne agg.InnerExceptions with
    | Some x -> unpackException x
    | None -> exn
  | _ -> exn
```

**Comment by bartelink** [comment]
> As long as it is a FSharp.Core issue I propose to use a corrected way of awaiting tasks https://github.com/jet/FsKafka/blob/6c6186f822a221adf0ad06e8a44a392aa55aa8ec/src/FsKafka/Infrastructure.fs#L14 instead of `Async.AwaitTask`

Note the edition therein in being synced with the canonical one in https://github.com/jet/FsKafka/pull/92 (similarly https://github.com/fsprojects/FSharp.AWS.DynamoDB/pull/49)

**Comment by bartelink** [comment]
@TheAngryByrd please excuse the atting....

what would you say about a breaking change regarding this?
I recently flipped some `taskResult` code to `asyncResult` and fell prey to the silent/implicit change in semantic (i.e. `with` no longer Doing What I Mean when awaiting something that returns `Task<'T>`
Possible resolutions are one or more of:
- make `asyncResult` not magically await tasks (force usage of `Async.AwaitTask` or an AwaitTaskCorrect of your choice)
- make the awaits of tasks in this lib undo the incorrect wrapping
- put AwaitTaskCorrect into this library and use it 

I guess the simplest thing is to remove implicit awaiting? (Checking if there's any way I can inhibit this as a consumer without changing the lib...)

**Comment by TheAngryByrd** [comment]
@bartelink 

I did create an “asyncEx” in IcedTasks with the unwrapping aggregate semantics. Could you give that a once over and make sure it’s working the way you think it should be? 

If it does then I think it might be worth making another CE that can polyfill the current one. 

Breaking API changes are less scary usually because there’s obvious compile errors but runtime changes does kind of scare me.

**Comment by bartelink** [comment]
Oh yeah, I remember that now 🤦
Will have a peep (there's another semantic in ATC about honoring cancellation too)
You're not wrong about compile vs runtime semantic changes being a very different level of concern

**Comment by bartelink** [comment]
Added a comment https://github.com/TheAngryByrd/IcedTasks/pull/24/files#r1556182882

I don't understand enough about what forces `AwaitAwaiter` is balancing to be able to comment on the specifics of its implementation

For me this boils down to two concerns
- having an implicit await behavior at all
- how many flavours there are in the world of that

For me, there can only be implicit behavior if there's a canonical expectation and/or spec for it. The canonical impl and the associated test snippets on fsssnip pin the behavior under cancellation too, and that's pretty critical for me

TL;DR I don't think any deviations whatsoever from `AwaitTaskCorrect` are a good idea.

I guess this tells us we need to get this into core because having TaskSeq, IcedTasks and/or here all have stuff that's not AwaitTaskCorrect is bad news (and I'd so love to stop having to cart it around github.com/jet and FsAwsDdb)

(Ironically, for my work project, having FsToolkit host AwaitTaskCorrect would be the perfect solution, but obviously having the world take a dependency on this entire lib for a correct awaittask impl is a step too far....)

**Comment by bartelink** [comment]
I think my conclusion/proposal is thus that the thing I'd do is remove the pieces that are bracketed with `#if !FABLE_COMPILER` for V5:

https://github.com/demystifyfp/FsToolkit.ErrorHandling/blob/dccdbbc629a15dbb1ab8c795ca515364cc81bfc7/src/FsToolkit.ErrorHandling/AsyncResultCE.fs#L201-L217
https://github.com/demystifyfp/FsToolkit.ErrorHandling/blob/dccdbbc629a15dbb1ab8c795ca515364cc81bfc7/src/FsToolkit.ErrorHandling/AsyncResultCE.fs#L131-L138

Luckily I'm not the maintainer of the library though, as I'd anticipate people immediately asking me: OK, so where it this wonderful `AwaitTaskCorrect` thing of which you speak so?

There's also the question of what to replace the usage of `Async.AwaitTask` in the `TryFinallyAsync` impl https://github.com/demystifyfp/FsToolkit.ErrorHandling/blob/dccdbbc629a15dbb1ab8c795ca515364cc81bfc7/src/FsToolkit.ErrorHandling/AsyncResultCE.fs#L57-L90

**Comment by bartelink** [comment]
@TheAngryByrd It would be a substantial sized PR, but I could stick AwaitTaskCorrect into the lib, and replace usage of AwaitTask with same for the v5 release line if you have the appetite

At the present time I dont see anyone out there with bandwidth for an AsyncEx lib (and my desire for this to be solved in a much smaller library to own this as any library cannot realistically pull in a library the size of this one as a dependency remains)

As hinted at above, the real answer is for the upstream support to get into FSharp.Core, but I guess ultimately there's going to be a need for two flavors in there, i.e. original AwaitTask with usage of AwaitTaskCorrect replacing that over time.

IMO at least having a common await function used internally would allow better solutions for async stacktraces etc to be experimented on over time.

Potentially a global lazy could allow one to opt out of AwaitTaskCorrect and have all awaits instead be plain AwaitTask

No pressure, but let me know if you have a view as to what's appropriate for v5

**Comment by bartelink** [comment]
FYI #373 will put an Async.Await shim that passes all FSharp.Core 11 equivalent tests in and use that
